kubetest/dind/dind.go

```diff
@@ -0,0 +1,397 @@
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+// Package dind implements dind-specific kubetest code.
+package dind
+
+import (
+	"fmt"
+	"io/ioutil"
+	"log"
+	"os"
+	"os/exec"
+	"strings"
+	"time"
+
+	"github.com/docker/docker/api/types"
+	"github.com/docker/docker/api/types/container"
+	"github.com/docker/docker/api/types/mount"
+	"github.com/docker/docker/client"
+	"golang.org/x/net/context"
+
+	"k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/tools/clientcmd"
+	"k8s.io/test-infra/kubetest/process"
+	"k8s.io/test-infra/kubetest/util"
+)
+
+type DindBuilder struct {
+	KubeRoot string
+	ToolRoot string
+	control  *process.Control
+}
+
+// NewBuilder returns an object capable of building the required k8s artifacts for dind.
+func NewBuilder(kubeRoot, toolRoot string, control *process.Control) *DindBuilder {
+	return &DindBuilder{
+		KubeRoot: kubeRoot,
+		ToolRoot: toolRoot,
+		control:  control,
+	}
+}
+
+// Build creates the k8s artifacts for dind.
+func (b *DindBuilder) Build() error {
+	// Set environment variables for the build, and go to our make dir. Then put everything back.
+	k, err := util.PushEnv("KUBE_ROOT", b.KubeRoot)
+	if err != nil {
+		return err
+	}
+	defer k()
+	t, err := util.PushEnv("TOOL_ROOT", b.ToolRoot)
+	if err != nil {
+		return err
+	}
+	defer t()
+	d, err := util.Pushd(b.ToolRoot)
+	if err != nil {
+		return err
+	}
+	defer d()
+
+	// Run the build command.
+	log.Printf("KUBE_ROOT: %s", os.Getenv("KUBE_ROOT"))
+	log.Printf("TOOL_ROOT: %s", os.Getenv("TOOL_ROOT"))
+	cmd := exec.Command("make", "-C", b.ToolRoot, "cluster")
+	return b.control.FinishRunning(cmd)
+}
+
+type DindTester struct {
+	kubecfg   string
+	ginkgo    string
+	e2etest   string
+	control   *process.Control
+	apiserver *kubernetes.Clientset
+	testArgs  *string
+	reportdir string
+}
+
+// NewTester returns an object that knows how to test the cluster it deployed.
+//TODO(Q-Lee): the deployer interfact should have a NewTester or Test method.
+func (d *DindDeployer) NewTester() (*DindTester, error) {
+	// Find the ginkgo and e2e.test artifacts we need. We'll cheat for now, and pull them from a known path.
+	// We only support dind from linux_amd64 anyway.
+	ginkgo := util.K8s("kubernetes", "bazel-bin", "vendor", "github.com", "onsi", "ginkgo", "ginkgo", "linux_amd64_stripped", "ginkgo")
+	e2etest := util.K8s("kubernetes", "bazel-bin", "test", "e2e", "e2e.test")
```

Comment:
do we want to ensure those two binary exists here?

Reply:
We don't strictly *need* to, because we'll just get the error later. But we can give a friendlier error message sooner if we do. The conformance test code I've got for you does that, so I'll make it consistent (that will make future refactors easier).